compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1204,6 +1257,122 @@ fn normalize_to_error<'a, 'tcx>(
+) -> Ty<'tcx> {
+    let tcx = selcx.tcx();
+
+    if !tcx.recursion_limit().value_within_limit(depth) {
+        // FIXME(fmease): Is there a better way to deal with overflow errors here? How can we detect
+        // cycles for example? Note that `report_overflow_{error,obligation_cycle}` are not quite
+        // suitable in this situation since they're tailored towards “requirements” (i.e. predicates).
+        // Of course, we could hackily construct an ad hoc `ProjectionPredicate` for it with RHS `_`.
+        //
+        // FIXME(fmease) Further, unless this is a cycle, suggest increasing the recursion limit.
+        //
+        // Halt compilation because it is important that overflows never be masked.
+        tcx.sess.emit_fatal(InherentProjectionNormalizationOverflow {
+            span: cause.span,
+            ty: alias_ty.to_string(),
+        });
+    }
+
+    let self_ty = alias_ty.self_ty();
+
+    // Confirm the inherent projection by registering the obligations
+    // arising from the corresponding impl.
+
+    let impl_ = tcx.parent(alias_ty.def_id);
+    let impl_substs = selcx.infcx.fresh_substs_for_item(cause.span, impl_);
+
+    let impl_bounds = tcx.predicates_of(impl_);
+    let impl_bounds = impl_bounds.instantiate(tcx, impl_substs);
+    let impl_bounds = normalize_with_depth_to(
+        selcx,
+        param_env,
+        cause.clone(),
+        depth + 1,
+        impl_bounds,
+        obligations,
+    );
+
+    let impl_obligations =
+        crate::traits::predicates_for_generics(|_, _| cause.clone(), param_env, impl_bounds);
+    obligations.extend(impl_obligations);
+
+    let impl_ty = tcx.type_of(impl_);
+    let impl_ty = impl_ty.subst(tcx, impl_substs);
+    let impl_ty =
+        normalize_with_depth_to(selcx, param_env, cause.clone(), depth + 1, impl_ty, obligations);
+
+    // Infer the generic parameters of the impl by unifying the
+    // impl type with the self type of the projection.
+    match selcx.infcx.at(&cause, param_env).eq(DefineOpaqueTypes::No, impl_ty, self_ty) {
+        Ok(mut ok) => obligations.append(&mut ok.obligations),
+        Err(_) => {
+            tcx.sess.delay_span_bug(
+                cause.span,
+                &format!(
+                    "{self_ty:?} was a subtype of {impl_ty:?} during selection but now it is not"
+                ),
+            );
+        }
+    }
+
+    let substs = alias_ty.rebase_substs_onto_impl(impl_substs, tcx);
+
+    // Register the obligations arising from the associated type itself.
+    {
+        let predicates = tcx.predicates_of(alias_ty.def_id).instantiate_own(tcx, substs);
+
+        for (predicate, span) in predicates {
+            let predicate = normalize_with_depth_to(
+                selcx,
+                param_env,
+                cause.clone(),
+                depth + 1,
+                predicate,
+                obligations,
+            );
+
+            let nested_cause = ObligationCause::new(
```

Comment:
Honestly you could probably just do this for all of the obligations -- removing `impl_bounds` above and instead calling `instantiate` on 1332.